mesonbuild/backend/backends.py

```diff
@@ -2046,3 +2053,10 @@ def compile_target_to_generator(self, target: build.CompileTarget) -> build.Gene
         all_sources = T.cast('_ALL_SOURCES_TYPE', target.sources) + T.cast('_ALL_SOURCES_TYPE', target.generated)
         return self.compiler_to_generator(target, target.compiler, all_sources,
                                           target.output_templ, target.depends)
+
+
+def compute_build_subdir(subdir: str, is_native_cross: bool) -> str:
+    if is_native_cross:
+        assert subdir.startswith('subprojects')
+        return 'subprojects-native' + subdir[11:]
+    return subdir
```

Comment:
Given that your target already knows that it's a build machine subproject, I think the approach of https://github.com/mesonbuild/meson/pull/12258/commits/4ead9fa8271f4ca5fffae4888c5b3a8ea419aae0 makes more sense, since it's part of the target itself. 

Essentially it splits the mirror assumption, and then provides `get_source_subdir()` and `get_output_subdir()` method to Target. No imports required.

Reply:
Thanks, that's much better indeed! Will try to cherry-pick that and retrofit on top of it.

Reply:
Done. I've also adapted 294d783e10a56e1422ba6d8623e19412e49bb25d.